pyomo/core/base/reference.py

```diff
@@ -410,7 +410,7 @@ def _identify_wildcard_sets(iter_stack, index):
         if len(index[i]) != len(level):
             return None
         # if any subset differs
-        if any(index[i].get(j,None) is not _set for j,_set in iteritems(level)):
+        if any(index[i].get(j,None) != _set for j,_set in iteritems(level)):
```

Comment:
Can you comment on the need for this change?  Th new code will be slower (`is` is an O(1) operation, where `!=` in general is O(n)).  Also, by not using `is`, something like the following:
```
m.b = Block([1,2])
m.I = Set(initialize=[1,2,3])
m.J = Set(initialize=[1,2,3])
m.b[1].x = Var(m.I)
m.b[2].x = Var(m.J)
s = m.b[:].x[:]
```
will identify wildcard sets for `s` as `[m.b_index, m.I]`, even though `m.b[2].x` is indexed by `m.J` and not `m.I`.  When we originally wrote this logic, the idea was to only return wildcard sets if the indexing sets were demonstrably "known".

Reply:
This fixes the toy model in issue #905 where a model containing a Block constructed using a rule like the one below had to have the discretization transformation come before the arc_expansion transformation.

```
m = ConcreteModel()

m.s = ContinuousSet(initialize=[0,10])

def block_rule(b, i):
    b.v = Var(['a','b'], initialize=1)
    return

m.b1 = Block(m.s, rule=block_rule)
m.b2 = Block(m.s, rule=block_rule)

m.p1 = Port()
m.p2 = Port()

m.r1 = Reference(m.b1[:].v[:])
```

The problem is that `b1[0].v` and `b1[10].v` are not identified as having the same indexing set because a new Set is being implicitly created from the values of the Python list. This causes the ContinuousSet index in the Reference component `r1` to be obscured so the indexed equality constraint that is eventually added by the arc_expansion transformation is not explicitly indexed by a ContinuousSet and therefore not expanded by the discretization transformation. 

I realize this change violates the original logic which requires the user to be much more explicit when declaring their indexing sets but I also think that the above implementation is pretty common and most users would expect `b1[0].v` and `b1[10].v` to be identified as having the same indexing set (they shouldn't have to know that a Set object is being created for them implicitly)

It's possible that there are edge cases here where if we were creating References for Params or Constraints (which aren't necessarily dense) then checking `is not` vs. `!=` leads to more or less intuitive behavior but for Vars I think it makes sense to check if all the elements of the sets are the same. 

A couple other ideas for solving this problem of transformation ordering are:

1. In the discretization transformation check component indexing sets for the `SetOf` type and automatically try to expand the component if it is found (since it might contain an obscured ContinuousSet) - _I'm worried this could lead to weird side-effects if the user does anything weird with adding elements to non-ContinuousSets_

2. Make References or SetOf aware of ContinuousSets - _Seems weird to make core code aware of something in Pyomo.DAE_

3. Modify the wildcard identification functionality to preserve any indexing sets that are identified as the same even if it encounters ones that aren't the same at a different level - _I'm not sure how hard this would be to implement_

Reply:
Let me add another option:
  4. Modify the wildcard identification so that it handles `SetOf differently (that is, use `is` for `Set` and `==` for `SetOf`).

That said, I think I like option 3.  Looking at the wildcard identification, I suspect that we *could* support a partial identification without too much hassle.  The challenge is that we would be potentially creating several pseudoset objects, and those objects would not be attached to the model.  Given that I am in favor of moving toward *not* requiring all set objects to be attached to the model (see #45), I don't see too much of a problem going down that path.  However, I think we can only handle "floating sets" correctly after we switch over to using the set rewrite (#326) by default.